### PR TITLE
Fix gelu returning NaN for infinite inputs on CPU and MPS

### PR DESCRIPTION
**Pull request by agarg0627**
#187293, #185770

If you call `F.gelu` on infinite inputs (for example, `F.gelu(torch.tensor([float("-inf")]))`), it doesn't return 0, which is the mathematical limit. Instead, it returns NaN. This is because `erf` saturates to -1 at `-inf`, so `0.5 * x * (1 + erf(x/sqrt(2)))` evaluates `-inf * 0`, which is NaN. The `tanh` approximation has the same problem. The reports also show `+inf` returning NaN through the vectorized `erf` on x86. The MPS Metal kernels share the formula, so they have the same bug.

I read through PR #185790, which was closed when its fork was deleted. Building on the diagnosis there, this PR adds infinity guards to the scalar and vectorized helpers in `Gelu.h` and to the two MPS functors, so `gelu(+inf)` returns `inf`, `gelu(-inf)` returns 0, and NaN propagation is unchanged. CUDA has the same bug, but I don't have a CUDA machine. As such, this PR covers CPU and MPS, and CUDA is left as a follow-up. The same guard is applied around the oneDNN gelu call in Activation.cpp, which handles contiguous CPU inputs on Linux, until this is fixed in oneDNN itself. This is independent of #189234, which reworks the same expression for a different bug and does not cover infinities.

Per review, the extreme values are covered through the OpInfo reference inputs instead of a standalone test, with expected failures for the tests that depend on the still-unfixed python ref or CUDA kernel.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @mikaylagawarecki

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -15630,6 +15630,21 @@ def test_RNN_input_size_zero(self, device):
             # Check that backward does not cause a hard error
             outs[0].sum().backward()
 
+    @skipCUDAIf(True, "gelu at infinity not yet fixed in the CUDA kernel")
+    @dtypes(torch.float32, torch.float64, torch.float16, torch.bfloat16)
+    @dtypesIfMPS(torch.float32, torch.float16, torch.bfloat16)
+    @parametrize_test("approximate", ["none", "tanh"])
+    def test_gelu_extreme_values(self, device, dtype, approximate):
```
Comment (Skylion007) [inline test/test_nn.py]: Why not just add these values as inputs to the opinfo tests?

Comment (agarg0627) [reply test/test_nn.py]: Adding infinities to the OpInfo samples would also test the python ref, the Inductor decomposition, and CUDA (via `test_compare_cpu`), which this PR doesn't fix, plus gradcheck at inf, which is undefined. I verified locally that the ref and grad tests fail with an infinite sample. I can move the coverage into the OpInfo in the CUDA follow-up!

Comment (kurtamohler) [reply test/test_nn.py]: It's probably best to just add it to opinfo and then add expected failures for any tests that don't support it

Comment (agarg0627) [reply test/test_nn.py]: Makes sense, just made those changes! I added the values via `reference_inputs_func` so they stay out of gradcheck, and marked expected failures for the ref-based tests.

File: torch/testing/_internal/common_methods_invocations.py
Comment (kurtamohler) [inline torch/testing/_internal/common_methods_invocations.py:17812]: `reference_gelu` will have to be updated to return the expected output for inf inputs to make `test_numpy_ref` pass for CPU and MPS.  Then you can also remove the test you added to `test_nn.py`

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -17793,6 +17803,9 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
                # May not replicate in CI
                DecorateInfo(unittest.skip("Skipped!"), 'TestCommon', 'test_out'),
                DecorateInfo(unittest.skip("Unsupported on MPS for now"), 'TestCommon', 'test_numpy_ref_mps'),
```
Comment (kurtamohler) [inline torch/testing/_internal/common_methods_invocations.py]: This MPS skip will need to be removed

File: torch/testing/_internal/common_methods_invocations.py
Comment (agarg0627) [reply torch/testing/_internal/common_methods_invocations.py:17812]: Done! Fixed `reference_gelu` for inf inputs and removed the `test_nn.py` test. Also, I had to add a cast back to the input dtype so `test_numpy_ref_mps` passes, because scipy's `norm.cdf` promotes float32 inputs to float64.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (agarg0627) [reply torch/testing/_internal/common_methods_invocations.py]: Yep got rid of it

**Comment by agarg0627** [comment]
@malfet @mruberry would one of you be able to take a look at this PR? happy to address any feedback, thank you!

**Comment by SrijanSuresh** [comment]
I have a CUDA machine (RTX 4060, sm_89) with a source build, so I am happy to take the CUDA follow-up once this lands: mirroring these guards in the CUDA kernel and removing the CUDA skip from the OpInfo. Let me know if that'd be useful.

One data point while I was reproducing on CPU (Zen 4, AVX-512, `2.15.0a0+git0f6199e`): `approximate="none"` returns `inf` for `+inf` at `n=1` but `NaN` at `n>=2`, while `approximate="tanh"` returns `inf` at every length. `-inf` gives `NaN` in both modes at every length. `torch.erf` itself returns `±1` correctly at all lengths on this machine, so whatever produces the `NaN` appears to be outside `erf` proper.

**Review by kurtamohler — APPROVED** [review]

**Comment by agarg0627** [comment]
@kurtamohler thanks for the approval! looks like workflows never got scheduled, would you be able to approve the workflows? thank you!

**Comment by kurtamohler** [comment]
Looks like at least one expectedFailure needs to be removed, and it also looks like the vmap output doesn't match

**Comment by agarg0627** [comment]
@kurtamohler Looks like on Linux, contiguous CPU inputs dispatch to oneDNN's gelu, which I missed, so eager still returns NaN there. The vmap mismatch is because batched inputs are non-contiguous and take the fixed native kernel. I was thinking of extending the fix to the oneDNN path (an inf fixup around the ideep call), or would you prefer a different approach?

**Comment by kurtamohler** [comment]
I'm not really sure what the best approach is, but I see three main options:

1. Try to get a PR through in mkldnn. Of course then we'd be waiting for the next release before being able to take advantage of it in pytorch, and it's also unknown right now if they would accept that change
2. Stop using mkldnn for gelu. I assume this would decrease performance significantly
3. Check for infs in the input and replace the nan outputs from mkldnn. This would also decrease performance, but maybe less so than option 2

Option 3 seems like the simplest way, and it might be worth checking to see how it affects performance. If it's only a small decrease then it might be fine

EDIT: Actually a possible 4th option might be that mkldnn's gelu already can do the nan replacement with a flag parameter or something. But I don't know if that's the case or not

**Comment by vpirogov** [comment]
> 1. Try to get a PR through in mkldnn. Of course then we'd be waiting for the next release before being able to take advantage of it in pytorch, and it's also unknown right now if they would accept that change

@kurtamohler, A PR would definitely be helpful. If this is indeed a bug we have an option to backport this change to oneDNN v3.13.2 and intercept PyTorch 2.15.

**Comment by agarg0627** [comment]
@vpirogov thank you! I'll file an issue in oneDNN with the repro and diagnosis.

@kurtamohler I checked option 4, but unfortunately there is no flag for this. I also implemented option 3 with a single-pass aminmax check on the input. On my Mac's oneDNN build, for finite inputs, there is 1-6% overhead for the erf mode and 14-18% for tanh (which only dispatches to oneDNN on aarch64). Would it be fine to push this as an interim and remove it once the oneDNN fix ships?

**Comment by kurtamohler** [comment]
> Would it be fine to push this as an interim and remove it once the oneDNN fix ships?

Potentially, but to make a good case for it, we'd probably have to have a performance measurement script that runs a good variety of different input shapes and a table of the results before and after the change

**Comment by agarg0627** [comment]
@kurtamohler Here is the script and measurements on my Mac:

<details><summary>benchmark script</summary>

```python
import sys
import torch
import torch.nn.functional as F
from torch.utils.benchmark import Timer

shapes = [
    (64,), (1024,), (65536,), (1048576,),
    (128, 768), (1024, 1024), (4096, 4096),
    (32, 128, 768), (8, 512, 1024), (32, 512, 4096),
    (16, 3, 224, 224), (64, 64, 56, 56),
]
label = sys.argv[1] if len(sys.argv) > 1 else "run"
print(f"# {label}: torch {torch.__version__}, threads={torch.get_num_threads()}, "
      f"mkldnn={torch.backends.mkldnn.is_available()}")
for approximate in ("none", "tanh"):
    for shape in shapes:
        x = torch.randn(*shape, dtype=torch.float32)
        t = Timer(
            "F.gelu(x, approximate=a)",
            globals={"F": F, "x": x, "a": approximate},
        ).blocked_autorange(min_run_time=3.0)
        print(f"{label},{approximate},{'x'.join(map(str, shape))},{t.median * 1e6:.2f}")
```
</details>

| mode | shape | before (us) | after (us) | change |
|---|---|---|---|---|
| none | 64 | 3.22 | 3.40 | +5.6% |
| none | 1024 | 9.93 | 10.23 | +3.0% |
| none | 65536 | 471.86 | 472.16 | +0.1% |
| none | 1048576 | 7524.83 | 7499.63 | -0.3% |
| none | 128x768 | 706.91 | 706.30 | -0.1% |
| none | 1024x1024 | 7513.48 | 7509.94 | -0.0% |
| none | 4096x4096 | 120211.79 | 119987.87 | -0.2% |
| none | 32x128x768 | 22519.98 | 22562.94 | +0.2% |
| none | 8x512x1024 | 30021.71 | 30119.44 | +0.3% |
| none | 32x512x4096 | 481302.67 | 482090.42 | +0.2% |
| none | 16x3x224x224 | 17276.50 | 17348.69 | +0.4% |
| none | 64x64x56x56 | 91869.71 | 92470.96 | +0.7% |
| tanh | 64 | 2.68 | 2.95 | +10.1% |
| tanh | 1024 | 3.30 | 3.69 | +11.8% |
| tanh | 65536 | 46.33 | 53.42 | +15.3% |
| tanh | 1048576 | 695.64 | 806.06 | +15.9% |
| tanh | 128x768 | 68.27 | 78.91 | +15.6% |
| tanh | 1024x1024 | 696.03 | 810.67 | +16.5% |
| tanh | 4096x4096 | 11093.38 | 13029.33 | +17.5% |
| tanh | 32x128x768 | 2082.02 | 2433.14 | +16.9% |
| tanh | 8x512x1024 | 2774.78 | 3238.66 | +16.7% |
| tanh | 32x512x4096 | 44370.65 | 51591.92 | +16.3% |
| tanh | 16x3x224x224 | 1594.60 | 1866.73 | +17.1% |
| tanh | 64x64x56x56 | 8492.94 | 9972.02 | +17.4% |

The tanh path only dispatches to oneDNN on aarch64, and the erf overhead is under 1% for all but tiny tensors.

**Comment by kurtamohler** [comment]
In my opinion, ~18% worst case slowdown is pretty inconsequential, especially if we'll remove the slowdown again in the long run. Let's do it

**Comment by agarg0627** [comment]
@kurtamohler great, I just pushed the changes. could you approve the workflows, thanks!

**Comment by agarg0627** [comment]
@Skylion007 would you be able to review these changes? I added the tests to OpInfo as requested. Thank you!
